datafusion/expr/src/logical_plan/builder.rs

```diff
@@ -188,37 +181,61 @@ impl LogicalPlanBuilder {
                     n_cols
                 );
             }
-            field_types = row
-                .iter()
-                .enumerate()
-                .map(|(j, expr)| {
-                    if let Expr::Literal(ScalarValue::Null) = expr {
-                        nulls.push((i, j));
-                        Ok(field_types[j].clone())
-                    } else {
-                        let data_type = expr.get_type(&empty_schema)?;
-                        if let Some(prev_data_type) = &field_types[j] {
-                            if prev_data_type != &data_type {
-                                return plan_err!("Inconsistent data type across values list at row {i} column {j}. Was {prev_data_type} but found {data_type}")
+        }
+
+        let empty_schema = DFSchema::empty();
```

Comment:
Another potential location for adding these coercion rules is in

https://github.com/apache/datafusion/blob/main/datafusion/optimizer/src/analyzer/type_coercion.rs

The benefit of putting them there is that it would work for plans created in ways other than the LogicalPlan builder. We could also potentially coerce types based on the target table's types rather than just the argument types themselves 🤔

Reply:
Hi @alamb, long time no see, thanks for the review. Do you mean to move the code for extracting the common types of values to `type_coercion.rs`? One problem is that `optimizer` crate depends on `expr` crate and we can't call the `optimizer` crate function in `LogicalPlanBuilder`. Maybe I'm not understanding you correctly, please give more advice to help me understand.

Reply:
Hi @b41sh 👋 

> . Do you mean to move the code for extracting the common types of values to type_coercion.rs? One problem is that optimizer crate depends on expr crate and we can't call the optimizer crate function in LogicalPlanBuilder. 

I was more thinking something like
1. Make no changes to `LogicalPlanBuilder` 
2. Instead, add code that coerces (adds casts) to the values in `LogicalPlan::Values` as part of the `TypeCoercion` rule

This would mean that the result of `LogicalPlanBuilder` would still have incompatible types until the `TypeCoercion` is run.

The benefit of this approach would be that when the `TypeCoercion` rule  is run, it would have the entire plan, including the table into which the values were being inserted, and so could maybe pick better types

The drawback is I think it would be a larger and more invasive change

Reply:
I understand what you mean, using the schema that the table already has can avoid redundant checking and casting, but there is a problem, `LogicalPlan::Values` needs a schema, and if we don't check the type at build time, we can't get a usable schema, which might be a bit problematic. Also, values can be executed separately without creating a table, for this scenario, it might be better to determine the schema at build time.